### PR TITLE
Draft - Introduce GH action running official dotnet formatter

### PR DESCRIPTION
**Pull request by Don-Vito**
Microsoft is developing .Net formatting  tool (https://github.com/dotnet/format)
that can (based on .editorconfig):
* enforce white spaces and indentation
* enforce code styling

There is also advance stuff, like "make readonly when possible".

My suggestion:
1. Define all our conventions including their severity in the .editorconfig
2. Run the formatter check as GitHub action 
(that can fail PR if issues with sev=warning are found)
3. Use .editorconfig as a default formatter in IDE (when applicable).

This PR contains:
1. A GitHub action that runs this tool
2. A sample configuration we can use with this action.

Why is it a draft? 
In addition to the importance to hear your feedbacks,
it is not  fully functional:  the tool succeeds to load only 
Files.Launcher and Common (and has fixes for them), 
but fails to load other project, complaining that
their dependencies cannot be found or misconfigured.

Exhausted with trying to fix them for today, so
dropping this here, in the case someone has suggestions,
opinions, knows how to make other projects load :smile:.

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+
+# Instance fields are camelCase and start with _
+dotnet_naming_rule.instance_fields_should_be_camel_case.severity = warning
+dotnet_naming_rule.instance_fields_should_be_camel_case.symbols = instance_fields
+dotnet_naming_rule.instance_fields_should_be_camel_case.style = instance_field_style
+
+dotnet_naming_symbols.instance_fields.applicable_kinds = field
+
+dotnet_naming_style.instance_field_style.capitalization = camel_case
+dotnet_naming_style.instance_field_style.required_prefix = _
```
Comment (tsvietOK) [inline .editorconfig:113]: All fields should be camelCase without underscore.

Comment (Don-Vito) [reply .editorconfig:113]: Cool. Thanks! I am glad you are reviewing this... there is much more beyond the styling (like use of var, etc.) and I have no clue what the community prefers.

**Comment by yaichenbaum** [comment]
@Don-Vito Does this tool auto format the code as well?

**Comment by Don-Vito** [comment]
> 
> 
> @Don-Vito Does this tool auto format the code as well?

For many inspections it can also format the code.
You can play with this locally - in the root directory copy the verision of .editconfig you like and run:
```
dotnet tool install -g dotnet-format --version 5.0.142902 --add-source https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet-tools/nuget/v3/index.json
dotnet-format --fix-style warn
```

**Comment by yaichenbaum** [comment]
@Don-Vito It seems to work well for detecting formatting issues, however, it doesn't seem to be able to fix any of them.

**Comment by Don-Vito** [comment]
> 
> 
> @Don-Vito It seems to work well for detecting formatting issues, however, it doesn't seem to be able to fix any of them.

Weird.. It fixed  some for me.. like removing the spaces, making attributes readonly, etc.
Were you able to run it on projects other than Files.Launcher and Common?

**Comment by Don-Vito** [comment]
BTW you can try to change --fix-style from warn to info, to make it more aggressive (and see more replacement capabilities)

**Comment by yaichenbaum** [comment]
> Were you able to run it on projects other than Files.Launcher and Common?

Nope, it didn't seem to find anything in those projects despite there being formating errors there.

> BTW you can try to change --fix-style from warn to info, to make it more aggressive (and see more replacement capabilities)

I gave this a try and it fixed some more issues, it looks like it needs to some more tweaking to fit our code styles. I will hand this over to @tsvietOK who has largly directed the code style for Files over the last couple of months.

**Comment by Don-Vito** [comment]
> 
> 
> > Were you able to run it on projects other than Files.Launcher and Common?
> 
> Nope, it didn't seem to find anything in those projects despite there being formating errors there.
> 
> > BTW you can try to change --fix-style from warn to info, to make it more aggressive (and see more replacement capabilities)
> 
> I gave this a try and it fixed some more issues, it looks like it needs to some more tweaking to fit our code styles. I will hand this over to @tsvietOK who has largly directed the code style for Files over the last couple of months.

@yair100 - it doesn't find anything in those projects as it fails to load them. In the projects it succeeds it seems to be doing a good job (I opened a simple project aside and it seems to kick asses). Will be glad to work on this with @tsvietOK.

**Comment by tsvietOK** [comment]
@Don-Vito When using command `dotnet-format --fix-style warn` or `dotnet-format --fix-style info` nothing changes for me.

**Comment by tsvietOK** [comment]
@Don-Vito For some reason `dotnet format` working only with Common and Files.Launcher projects. `dotnet restore` doesn't help.

**Comment by Don-Vito** [comment]
> 
> 
> @Don-Vito For some reason `dotnet format` working only with Common and Files.Launcher projects. `dotnet restore` doesn't help.

@tsvietOK - yes. This what I am trying to debug. When I play with Files.Launcher the tool seems to do a great job. But for other projects it simply fails and dotnet restore (that I added to the action as well) doesn't help. I will try to open an issue in their tracker.

**Comment by heistema** [comment]
If I understood correctly we are talking about whether "dotnet-format" is capable of enforcing our needed set of ".editorconfig" rules. Which rules we need and whether the supported subset of "dotnet-format" rules is sufficient for our needs is another question (another issue?).

I think I would be fine with both tools... The question I have is whether the integration in Visual Studio is a bit better with "NetAnalyzers". Also multi-project support is not that crucial with "NetAnalyzers" as long as you have the .editorconfig in the top level of code hierarchy (Solution files).

This is only my impression after having a short test with both of them.

**Pull request closed without merging**